packages/material-ui/src/Menu/Menu.js

```diff
@@ -72,6 +78,58 @@ const Menu = React.forwardRef(function Menu(props, ref) {
     }
   };
 
+  let firstValidElementIndex = null;
+  let firstFocusableIndex = null;
+  let firstAutoFocusIndex = null;
+  let firstSelectedIndex = null;
+  React.Children.forEach(children, (child, index) => {
```

Comment:
Do you see an opportunity for not relying on children iteration?

Reply:
It would be possible for the `menu` variant, but for the others I couldn't think of any robust approach that doesn't require knowing whether or not we have a selected child later in the children in order to know how to treat the first child (i.e. whether or not it should be the content anchor and/or auto-focused).

Previously `Menu` relied on `MenuList` figuring this out during its `React.Children.map` iteration (which is now gone), but I think it will be much easier to further enhance `Menu` going forward with it knowing/managing this information.

I do think I can simplify this a little and get rid of the `firstAutoFocusIndex` variable. My thought with having it, was for it to work more robustly by default if someone uses `autoFocus` explicitly on an item, but I think the user should probably just set Menu `autoFocus` to false if they are trying to manage focus of particular items rather than `Menu` bothering with trying to look for this case.

Still to be added in this area is checking for disabled items and dividers so that we don't apply `autoFocus` to them.

Reply:
After further thought, I'm pretty sure I can get rid of the `forEach` and just have the one `map` of the children if I make a slight change to the SelectInput behavior. It would actually remove the need for the `select` variant and `SelectInput` would just use the default `selectedMenu` variant.

Currently, if SelectInput does not have a selected item, then focus is given to the first item. However, there isn't a compelling reason to do this. In a native select, the first item is automatically selected if no other index is explicitly selected, but that isn't the case for SelectInput which requires a value and the value fully controls which item is selected.

Since the value is required for SelectInput, the only time when nothing is selected is if there isn't a menu item with a matching value. This is certainly a possible case, but it is not a case that represents recommended usage; and it is arguably just as good or better behavior for this case to behave the same as the other variants and put focus on `MenuList` in that case rather than the first item so that we don't give the false impression that the first item is already selected.

With that change and the autoFocus simplification I already mentioned, I'm pretty sure I can achieve the rest of the functionality in a single pass (though probably requiring an additional ref to help figure out the content anchor accurately).

It will probably be a few days before I can address the review items. None of them sound like big changes, I just probably won't have much time to spend on it prior to the weekend.